File: src/main/java/org/example/kalkulationsprogramm/service/KalenderService.java
package org.example.kalkulationsprogramm.service;

import lombok.RequiredArgsConstructor;
import org.example.kalkulationsprogramm.domain.*;
import org.example.kalkulationsprogramm.repository.*;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.time.YearMonth;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

@Service
@RequiredArgsConstructor
public class KalenderService {

    private final KalenderEintragRepository kalenderEintragRepository;
    private final ProjektRepository projektRepository;
    private final KundeRepository kundeRepository;
    private final LieferantenRepository lieferantenRepository;
    private final AngebotRepository angebotRepository;
    private final MitarbeiterRepository mitarbeiterRepository;

    /**
     * Lädt alle Kalendereinträge für einen bestimmten Monat (PC-Ansicht, alle Einträge).
     */
    public List<KalenderEintrag> getEintraegeForMonat(int jahr, int monat) {
        YearMonth yearMonth = YearMonth.of(jahr, monat);
        LocalDate von = yearMonth.atDay(1);
        LocalDate bis = yearMonth.atEndOfMonth();
        return kalenderEintragRepository.findByDatumBetween(von, bis);
    }

    /**
     * Lädt alle Kalendereinträge für einen Mitarbeiter in einem Monat.
     * Zeigt Termine wo Mitarbeiter Ersteller oder Teilnehmer ist, plus Firmenkalender-Einträge.
     */
    public List<KalenderEintrag> getEintraegeForMitarbeiter(Long mitarbeiterId, int jahr, int monat) {
        YearMonth yearMonth = YearMonth.of(jahr, monat);
        LocalDate von = yearMonth.atDay(1);
        LocalDate bis = yearMonth.atEndOfMonth();
        return kalenderEintragRepository.findByMitarbeiterAndDatumBetween(mitarbeiterId, von, bis);
    }

    /**
     * Lädt alle Kalendereinträge für einen Mitarbeiter an einem bestimmten Tag.
     */
    public List<KalenderEintrag> getEintraegeForMitarbeiterTag(Long mitarbeiterId, LocalDate datum) {
        return kalenderEintragRepository.findByMitarbeiterAndDatum(mitarbeiterId, datum);
    }

    /**
     * Lädt alle Kalendereinträge für einen Datumsbereich.
     */
    public List<KalenderEintrag> getEintraegeForRange(LocalDate von, LocalDate bis) {
        return kalenderEintragRepository.findByDatumBetween(von, bis);
    }

    /**
     * Lädt einen einzelnen Eintrag.
     */
    public KalenderEintrag getEintrag(Long id) {
        return kalenderEintragRepository.findById(id).orElse(null);
    }

    /**
     * Lädt einen einzelnen Eintrag mit Teilnehmern.
     */
    public KalenderEintrag getEintragWithTeilnehmer(Long id) {
        return kalenderEintragRepository.findByIdWithTeilnehmer(id);
    }

    /**
     * Speichert einen Kalendereintrag (neu oder aktualisiert).
     * Unterstützt jetzt auch Ersteller und Teilnehmer.
     */
    @Transactional
    public KalenderEintrag saveEintrag(KalenderEintrag eintrag, Long projektId, Long kundeId, Long lieferantId,
            Long angebotId, Long erstellerId, List<Long> teilnehmerIds) {
        // Verknüpfungen setzen
        if (projektId != null) {
            eintrag.setProjekt(projektRepository.findById(projektId).orElse(null));
        } else {
            eintrag.setProjekt(null);
        }

        if (kundeId != null) {
            eintrag.setKunde(kundeRepository.findById(kundeId).orElse(null));
        } else {
            eintrag.setKunde(null);
        }

        if (lieferantId != null) {
            eintrag.setLieferant(lieferantenRepository.findById(lieferantId).orElse(null));
        } else {
            eintrag.setLieferant(null);
        }

        if (angebotId != null) {
            eintrag.setAngebot(angebotRepository.findById(angebotId).orElse(null));
        } else {
            eintrag.setAngebot(null);
        }

        // Ersteller setzen (nur bei neuem Eintrag, wenn noch kein Ersteller)
        if (erstellerId != null && eintrag.getErsteller() == null) {
            eintrag.setErsteller(mitarbeiterRepository.findById(erstellerId).orElse(null));
        }

        // Teilnehmer setzen
        if (teilnehmerIds != null) {
            Set<Mitarbeiter> teilnehmer = new HashSet<>();
            for (Long teilnehmerId : teilnehmerIds) {
                mitarbeiterRepository.findById(teilnehmerId).ifPresent(teilnehmer::add);
            }
            eintrag.setTeilnehmer(teilnehmer);
        }

        return kalenderEintragRepository.save(eintrag);
    }

    /**
     * Legacy-Methode für Kompatibilität ohne Ersteller/Teilnehmer.
     */
    @Transactional
    public KalenderEintrag saveEintrag(KalenderEintrag eintrag, Long projektId, Long kundeId, Long lieferantId,
            Long angebotId) {
        return saveEintrag(eintrag, projektId, kundeId, lieferantId, angebotId, null, null);
    }

    /**
     * Löscht einen Kalendereintrag.
     */
    @Transactional
    public void deleteEintrag(Long id) {
        kalenderEintragRepository.deleteById(id);
    }

    /**
     * Lädt Einträge für ein bestimmtes Projekt.
     */
    public List<KalenderEintrag> getEintraegeForProjekt(Long projektId) {
        return kalenderEintragRepository.findByProjektIdOrderByDatumDesc(projektId);
    }
}

File: src/main/kotlin/org/example/kalkulationsprogramm/service/KalenderService.kt
package org.example.kalkulationsprogramm.service

import org.example.kalkulationsprogramm.domain.KalenderEintrag
import org.example.kalkulationsprogramm.domain.Mitarbeiter
import org.example.kalkulationsprogramm.repository.AnfrageRepository
import org.example.kalkulationsprogramm.repository.KalenderEintragRepository
import org.example.kalkulationsprogramm.repository.KundeRepository
import org.example.kalkulationsprogramm.repository.LieferantenRepository
import org.example.kalkulationsprogramm.repository.MitarbeiterRepository
import org.example.kalkulationsprogramm.repository.ProjektRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDate
import java.time.YearMonth

@Service
open class KalenderService(
    private val kalenderEintragRepository: KalenderEintragRepository,
    private val projektRepository: ProjektRepository,
    private val kundeRepository: KundeRepository,
    private val lieferantenRepository: LieferantenRepository,
    private val anfrageRepository: AnfrageRepository,
    private val mitarbeiterRepository: MitarbeiterRepository,
) {

    open fun getEintraegeForMonat(jahr: Int, monat: Int): List<KalenderEintrag> {
        val yearMonth = YearMonth.of(jahr, monat)
        return kalenderEintragRepository.findByDatumBetween(yearMonth.atDay(1), yearMonth.atEndOfMonth())
    }

    open fun getEintraegeForMitarbeiter(mitarbeiterId: Long?, jahr: Int, monat: Int): List<KalenderEintrag> {
        val yearMonth = YearMonth.of(jahr, monat)
        return kalenderEintragRepository.findByMitarbeiterAndDatumBetween(
            mitarbeiterId,
            yearMonth.atDay(1),
            yearMonth.atEndOfMonth(),
        )
    }

    open fun getEintraegeForMitarbeiterTag(mitarbeiterId: Long?, datum: LocalDate): List<KalenderEintrag> =
        kalenderEintragRepository.findByMitarbeiterAndDatum(mitarbeiterId, datum)

    open fun getEintraegeForRange(von: LocalDate, bis: LocalDate): List<KalenderEintrag> =
        kalenderEintragRepository.findByDatumBetween(von, bis)

    open fun getEintrag(id: Long?): KalenderEintrag? =
        kalenderEintragRepository.findById(id!!).orElse(null)

    open fun getEintragWithTeilnehmer(id: Long?): KalenderEintrag =
        kalenderEintragRepository.findByIdWithTeilnehmer(id)

    @Transactional
    open fun saveEintrag(
        eintrag: KalenderEintrag,
        projektId: Long?,
        kundeId: Long?,
        lieferantId: Long?,
        anfrageId: Long?,
        erstellerId: Long?,
        teilnehmerIds: List<Long>?,
    ): KalenderEintrag {
        eintrag.projekt = projektId?.let { projektRepository.findById(it).orElse(null) }
        eintrag.kunde = kundeId?.let { kundeRepository.findById(it).orElse(null) }
        eintrag.lieferant = lieferantId?.let { lieferantenRepository.findById(it).orElse(null) }
        eintrag.anfrage = anfrageId?.let { anfrageRepository.findById(it).orElse(null) }

        if (erstellerId != null && eintrag.ersteller == null) {
            eintrag.ersteller = mitarbeiterRepository.findById(erstellerId).orElse(null)
        }

        if (teilnehmerIds != null) {
            val teilnehmer = mutableSetOf<Mitarbeiter>()
            for (teilnehmerId in teilnehmerIds) {
                mitarbeiterRepository.findById(teilnehmerId).ifPresent { teilnehmer.add(it) }
            }
            eintrag.teilnehmer = teilnehmer
        }

        return kalenderEintragRepository.save(eintrag)
    }

    @Transactional
    open fun saveEintrag(
        eintrag: KalenderEintrag,
        projektId: Long?,
        kundeId: Long?,
        lieferantId: Long?,
        anfrageId: Long?,
    ): KalenderEintrag =
        saveEintrag(eintrag, projektId, kundeId, lieferantId, anfrageId, null, null)

    @Transactional
    open fun deleteEintrag(id: Long?) {
        kalenderEintragRepository.deleteById(id!!)
    }

    open fun getEintraegeForProjekt(projektId: Long?): List<KalenderEintrag> =
        kalenderEintragRepository.findByProjektIdOrderByDatumDesc(projektId)
}
